src/java.base/share/classes/jdk/internal/classfile/impl/RawBytecodeHelper.java

```diff
@@ -51,121 +94,198 @@ public static int align(int n) {
-                    if (aligned_bci + 2 * 4 >= endBci) {
-                        yield -1;
-                    }
-                    int npairs = bytecode.getInt(aligned_bci + 4);
-                    int l = aligned_bci - bci + (2 + 2 * npairs) * 4;
-                    if (l > 0) yield l; else yield -1;
-                }
-                default ->
-                    0;
-            };
-            if (len <= 0 || (bci > endBci - len) || (bci - len >= nextBci)) {
-                code = ILLEGAL;
-            } else {
-                nextBci += len;
-                isWide = false;
-                if (code == WIDE) {
-                    if (bci + 1 >= endBci) {
-                        code = ILLEGAL;
-                    } else {
-                        code = bytecode.get(bci + 1) & 0xff;
-                        isWide = true;
-                    }
-                }
-            }
-            rawCode = code;
-            return code;
-        }
+    public int getShort(int bci) {
+        Preconditions.checkFromIndexSize(bci, 2, endBci(), IAE_FORMATTER);
+        return getShortUnchecked(bci);
+    }
+
+    public int getInt(int bci) {
+        Preconditions.checkFromIndexSize(bci, 4, endBci(), IAE_FORMATTER);
+        return getIntUnchecked(bci);
+    }
+
+    // Unchecked accessors: only if opcode() is validated
+
+    public int getU1Unchecked(int bci) {
+        return Byte.toUnsignedInt(code.array[bci]);
+    }
+
+    public int getU2Unchecked(int bci) {
+        return UNSAFE.getCharUnaligned(code.array, (long) Unsafe.ARRAY_BYTE_BASE_OFFSET + bci, true);
     }
 
+    public int getShortUnchecked(int bci) {
+        return UNSAFE.getShortUnaligned(code.array, (long) Unsafe.ARRAY_BYTE_BASE_OFFSET + bci, true);
+    }
+
+    // used after switch validation
+    public int getIntUnchecked(int bci) {
+        return UNSAFE.getIntUnaligned(code.array, (long) Unsafe.ARRAY_BYTE_BASE_OFFSET + bci, true);
+    }
+
+    // non-wide branches
+    public int dest() {
+        return bci + getShortUnchecked(bci + 1);
+    }
+
+    // goto_w and jsr_w
+    public int destW() {
+        return bci + getIntUnchecked(bci + 1);
+    }
+
+    // *load, *store, iinc
     public int getIndex() {
-        return (isWide) ? getIndexU2Raw(bci + 2) : getIndexU1();
+        return isWide ? getU2Unchecked(bci + 2) : getIndexU1();
     }
 
+    // ldc
+    public int getIndexU1() {
+        return getU1Unchecked(bci + 1);
+    }
+
+    // usually cp entry index
     public int getIndexU2() {
-        return getIndexU2Raw(bci + 1);
+        return getU2Unchecked(bci + 1);
+    }
+
+    // Transition methods
+
+    /**
+     * Transitions to the next instruction and returns whether scanning should
+     * continue. If the next instruction is malformed, {@link #opcode()} returns
+     * {@link #ILLEGAL}, so we can perform value access without bound checks if
+     * we have a valid opcode.
+     */
+    public boolean next() {
```

Comment:
In C1, this cannot be inlined. See if you need to add ForceInline

Reply:
I don't think we should worry too much about making C1 inline more aggressively.

Reply:
C2 needs 10000 calls to inline this method, so wenshao is worried. However, this method call is almost always followed by a huge switch to handle different opcode, so I doubt how much of a difference inlining brings.

But wenshao discovered that this method has too many field gets; I should indeed convert them to local variable access if possible.

Reply:
Yes, things like storing `endBci()` to a local variable can be great if it both reduces code size and improves interpreter/C1 speed - but don't over-do it as it's likely never-ending work for a kind of optimizations leyden might make pointless.

Reply:
It is not necessary to use RawBytecodeHelper in the StackMapGenerator#detectFrameOffsets method. It may be better to operate directly on the `byte[] bytecode`.

Reply:
We could have just tracked the offsets when labels from DirectCodeBuilder are bound; we can explore such a fix in another patch.